File: apps/android-twa/app/src/main/java/com/agentic/wallet/agent/provider/AnthropicProvider.kt
package com.agentic.wallet.agent.provider

import com.agentic.wallet.agent.prompts.DeviceAgentMessageAssembler
import com.agentic.wallet.agent.prompts.Messages
import com.agentic.wallet.agent.runtime.RuntimeConfig
import org.json.JSONArray
import org.json.JSONObject
import java.time.Clock
import java.time.Instant

// Anthropic Device Agent provider — POST /v1/messages with optional web_search tool.
// Two-pass research flow: when research.needed=true, runs a free-text research call with
// web_search bound (no JSON constraint), captures filtered citations + summary, then runs
// a structured review pass with the research evidence injected into context.
// Mirrors apps/browser-demo/src/deviceAgent/provider/anthropicProvider.ts (which omits the
// browser CORS header — Android's HttpURLConnection doesn't need it).
internal class AnthropicProvider(
    private val config: RuntimeConfig,
    private val http: HttpExecutor,
    private val clock: Clock = Clock.systemUTC(),
) : DeviceAgentProvider {

    override suspend fun generatePlan(payload: JSONObject): JSONObject {
        val messages = DeviceAgentMessageAssembler.buildPlanMessages(payload)
        val response = postMessages(messages, PLAN_MAX_TOKENS, PLAN_TEMPERATURE, payload)
        return ProviderResponseParser.parseModelJson(ProviderResponseParser.extractAnthropicText(response))
    }

    override suspend fun reviewPlan(payload: JSONObject): JSONObject {
        // Two-pass flow for parity with the local-bridge planner: when the review needs outside
        // facts, run a research-only call (web_search bound, no JSON requirement) first, then a
        // structured-output call with the research summary embedded in context. Avoids the model
        // juggling "search the web" + "return JSON" at the same time (Helium NOTE single-pass
        // returned $20 — a matching plan — vs $15 from the two-pass local-bridge).
        if (researchNeeded(payload)) {
            val enriched = runResearchPass(payload)
            val reviewPayload = mergeResearchSignal(enriched)
            val messages = DeviceAgentMessageAssembler.buildReviewMessages(reviewPayload)
            val response = postMessages(messages, REVIEW_MAX_TOKENS, REVIEW_TEMPERATURE, reviewPayload)
            return ReviewPostprocessor.finalize(
                ProviderResponseParser.parseModelJson(ProviderResponseParser.extractAnthropicText(response)),
                reviewPayload,
            )
        }
        val messages = DeviceAgentMessageAssembler.buildReviewMessages(payload)
        val response = postMessages(messages, REVIEW_MAX_TOKENS, REVIEW_TEMPERATURE, payload)
        return ReviewPostprocessor.finalize(
            ProviderResponseParser.parseModelJson(ProviderResponseParser.extractAnthropicText(response)),
            payload,
        )
    }

    override suspend fun ask(payload: JSONObject): JSONObject {
        val messages = DeviceAgentMessageAssembler.buildAskMessages(payload)
        val response = postMessages(messages, ASK_MAX_TOKENS, ASK_TEMPERATURE, payload)
        val text = ProviderResponseParser.extractAnthropicText(response)
        if (text.isBlank()) {
            throw ProviderHttpException(ProviderErrorCodes.INVALID_RESPONSE, "Provider response had no answer text.")
        }
        return JSONObject().put("output_text", text)
    }

    /**
     * Research pass — separate LLM call with web search bound. Captures the model's research
     * summary + citations and returns the original payload with `context.researchEvidence`
     * populated, ready for the structured review pass to consume. Failures are non-fatal.
     */
    private suspend fun runResearchPass(payload: JSONObject): JSONObject {
        return try {
            val messages = DeviceAgentMessageAssembler.buildResearchMessages(payload, researchTargetsForPayload(payload))
            // Force research.needed=true on the inner payload so postMessages attaches web_search.
            val innerPayload = copyJson(payload)
            val innerResearch = innerPayload.optJSONObject("research") ?: JSONObject()
            innerResearch.put("needed", true)
            innerPayload.put("research", innerResearch)

            val response = postMessages(messages, REVIEW_MAX_TOKENS, REVIEW_TEMPERATURE, innerPayload)
            val rawSummary = ProviderResponseParser.extractAnthropicText(response).trim()
            val rawCitations = ProviderResponseParser.extractAnthropicCitations(response)
            val instruction = extractInstructionText(payload)
            val filteredCitations = CitationFilter.filterLowAuthorityCitations(rawCitations, instruction)

            // A pricing question with no usable official citation is unverified — whether citations
            // were filtered out as low-authority OR the provider returned none at all (a model
            // answering from training because its web-search tool silently never ran, the
            // OpenRouter+Claude Helium "$0"). Never propagate an un-sourced price; force needs_input.
            val pricingQuestion = CitationFilter.isPricingInstruction(instruction)
            val unverifiedPricing = pricingQuestion && filteredCitations.isEmpty()
            val summary = if (unverifiedPricing) {
                "Current pricing could not be verified against an official source. Ask the user to confirm the plan name and price."
            } else {
                rawSummary.ifEmpty { "Research ran but produced no summary text." }
            }

            if (filteredCitations.isEmpty() && rawSummary.isEmpty()) return payload

            val sources = JSONArray()
            for (c in filteredCitations) {
                val src = JSONObject().put("url", c.url)
                if (c.title != null) src.put("title", c.title)
                sources.put(src)
            }
            val researchEvidence = JSONObject()
                .put("status", "checked")
                .put("required", true)
                .put("provider", "Anthropic")
                .put("checkedAt", Instant.now(clock).toString())
                .put("summary", summary)
                .put("sources", sources)
                .put(
                    "sourcePolicy",
                    "Prefer official sources for prices and product facts. When a vendor publishes a plan page, use it as primary. Reject blog subdomains (blog.*, news.*) as primary sources for current pricing. Cite each fact with the official URL.",
                )

            val merged = copyJson(payload)
            val mergedContext = merged.optJSONObject("context") ?: JSONObject()
            mergedContext.put("researchEvidence", researchEvidence)
            merged.put("context", mergedContext)
            merged
        } catch (cancel: kotlinx.coroutines.CancellationException) {
            throw cancel
        } catch (_: Throwable) {
            // Research pass failure must not block the review — return the payload unchanged.
            payload
        }
    }

    private suspend fun postMessages(
        messages: Messages,
        maxTokens: Int,
        temperature: Double,
        payload: JSONObject,
    ): JSONObject {
        val apiKey = (config.apiKey ?: "").trim()
        ProviderHttp.assertApiKeyHeaderSafe(apiKey)
        val apiFormat = if (isOpenRouterConfig()) "openai-compatible" else "anthropic"
        val baseUrl = ProviderHttp.normalizeBaseUrl(config.baseUrl, apiFormat)
        val url = "$baseUrl/messages"
        val body = buildRequestBody(messages, maxTokens, temperature, payload)
        val headers = if (isOpenRouterConfig()) {
            mapOf(
                "Authorization" to "Bearer $apiKey",
                "X-OpenRouter-Metadata" to "enabled",
            ) + ProviderHttp.openRouterAttributionHeaders(true)
        } else {
            mapOf(
                "x-api-key" to apiKey,
                "anthropic-version" to ANTHROPIC_VERSION,
            )
        }
        val response = http.postJson(url, headers, body.toString())
        val errorCode = ProviderHttp.mapHttpStatusToErrorCode(response.status)
        if (errorCode != null) {
            throw ProviderHttpException(
                errorCode,
                ProviderHttp.composeErrorMessage(response.status, response.body),
            )
        }
        return try {
            JSONObject(response.body)
        } catch (_: Throwable) {
            throw ProviderHttpException(
                ProviderErrorCodes.INVALID_RESPONSE,
                "Provider response was not valid JSON.",
            )
        }
    }

    private fun buildRequestBody(
        messages: Messages,
        maxTokens: Int,
        temperature: Double,
        payload: JSONObject,
    ): JSONObject {
        val userMessages = JSONArray().put(
            JSONObject()
                .put("role", "user")
                .put("content", messages.userContent),
        )
        val body = JSONObject()
            .put("model", config.model.trim())
            .put("max_tokens", maxTokens)
            .put("system", messages.system)
            .put("messages", userMessages)
            .put("temperature", temperature)
        if (researchNeeded(payload)) {
            body.put(
                "tools",
                JSONArray().put(
                    webSearchTool(payload),
                ),
            )
        }
        return body
    }

    private fun isOpenRouterConfig(): Boolean =
        config.provider.trim().equals("openrouter", ignoreCase = true) ||
            (config.baseUrl ?: "").contains("openrouter.ai", ignoreCase = true)

    // This provider always speaks the Anthropic Messages format (/messages), whether direct or
    // via OpenRouter's Anthropic-compat skin. OpenRouter's openrouter:web_search server tool only
    // works on its Chat Completions / Responses endpoints — NOT the Messages skin — so binding it
    // here meant the tool was silently dropped and Claude answered ungrounded (the
    // OpenRouter+Claude Helium "$0" bug). Always bind Anthropic's NATIVE web_search tool;
    // OpenRouter's skin forwards native tool use to Anthropic 1P.
    private fun webSearchTool(payload: JSONObject): JSONObject =
        JSONObject()
            .put("type", "web_search_20250305")
            .put("name", "web_search")
            .put("max_uses", researchMaxUses(payload))
            .put(
                "user_location",
                JSONObject()
                    .put("type", "approximate")
                    .put("country", "US")
                    .put("timezone", "America/Los_Angeles"),
            )

    companion object {
        private const val ANTHROPIC_VERSION: String = "2023-06-01"
        private const val PLAN_MAX_TOKENS: Int = 1024
        private const val REVIEW_MAX_TOKENS: Int = 1800
        private const val ASK_MAX_TOKENS: Int = 800
        private const val PLAN_TEMPERATURE: Double = 0.2
        private const val REVIEW_TEMPERATURE: Double = 0.2
        private const val ASK_TEMPERATURE: Double = 0.3

        private fun researchNeeded(payload: JSONObject): Boolean =
            payload.optJSONObject("research")?.optBoolean("needed", false) == true

        private fun researchMaxUses(payload: JSONObject): Int {
            val raw = payload.optJSONObject("research")?.opt("maxSearches")
            val numeric = when (raw) {
                is Number -> raw.toInt()
                else -> 3
            }
            return numeric.coerceIn(1, 5)
        }

        private fun extractInstructionText(payload: JSONObject): String {
            val direct = payload.optString("instruction", "")
            if (direct.isNotEmpty()) return direct
            val userPrompt = payload.optString("userPrompt", "")
            if (userPrompt.isNotEmpty()) return userPrompt
            return payload.optString("question", "")
        }

        private fun mergeResearchSignal(payload: JSONObject): JSONObject {
            val merged = copyJson(payload)
            val existing = merged.optJSONObject("research") ?: JSONObject()
            existing.put("needed", false)
            existing.put("mode", "provided_current_facts")
            existing.put("providedEvidence", true)
            merged.put("research", existing)
            return merged
        }

        private fun copyJson(src: JSONObject): JSONObject = JSONObject(src.toString())
    }
}

File: apps/browser-demo/src/deviceAgent/provider/anthropicProvider.ts
// TypeScript port of AnthropicProvider.kt with one deliberate divergence:
// the `anthropic-dangerous-direct-browser-access: 'true'` header. Browsers
// without this header are blocked by Anthropic's CORS policy; Android's
// HttpURLConnection has no CORS, so the Kotlin runtime omits it (and its
// AnthropicProviderTest pins the absence). Mirrors the planner.ts pattern.

import { buildAskMessages, buildPlanMessages, buildResearchMessages, buildReviewMessages } from '../prompts/messageAssembler.js';
import type { DeviceAgentMessages } from '../prompts/messageAssembler.js';
import type { RuntimeConfig } from '../runtime/config.js';

import { filterLowAuthorityCitations, isPricingInstruction } from './citationFilter.js';
import { PROVIDER_ERROR_CODES, ProviderHttpError } from './errorCodes.js';
import type { HttpExecutor } from './http.js';
import { openRouterAttributionHeaders } from './openRouterHeaders.js';
import {
  assertApiKeyHeaderSafe,
  composeErrorMessage,
  mapHttpStatusToErrorCode,
  normalizeBaseUrl,
} from './providerHttp.js';
import { researchTargetsForPayload } from './researchTargets.js';
import { extractAnthropicCitations, extractAnthropicText, parseModelJson } from './responseParser.js';
import { finalizeReviewResultForPayload } from './reviewPostprocess.js';
import type { DeviceAgentProvider } from './types.js';

const ANTHROPIC_VERSION = '2023-06-01';
const PLAN_TEMPERATURE = 0.2;
const REVIEW_TEMPERATURE = 0.2;
const ASK_TEMPERATURE = 0.3;
const PLAN_MAX_TOKENS = 1024;
const REVIEW_MAX_TOKENS = 1800;
const ASK_MAX_TOKENS = 800;

export class AnthropicProvider implements DeviceAgentProvider {
  private readonly config: RuntimeConfig;
  private readonly http: HttpExecutor;

  constructor(config: RuntimeConfig, http: HttpExecutor) {
    this.config = config;
    this.http = http;
  }

  async generatePlan(payload: Record<string, unknown>, signal?: AbortSignal): Promise<Record<string, unknown>> {
    const messages = buildPlanMessages(payload);
    const response = await this.postMessages(messages, PLAN_MAX_TOKENS, PLAN_TEMPERATURE, signal, payload);
    return parseModelJson(extractAnthropicText(response));
  }

  async reviewPlan(payload: Record<string, unknown>, signal?: AbortSignal): Promise<Record<string, unknown>> {
    // Two-pass flow for parity with the local-bridge planner: when the review needs
    // outside facts, run a research-only call (web_search bound, no JSON requirement)
    // first, then a structured-output call with the research summary embedded in context.
    // This avoids the model juggling "search the web" + "return JSON" at the same time,
    // which was the failure mode on the Helium NOTE (Device Agent single-pass returned
    // $20 — a matching plan — while local-bridge two-pass returned $15, the cheapest).
    if (researchNeeded(payload)) {
      const enrichedPayload = await this.runResearchPass(payload, signal);
      // Second pass: structured review, NO web_search bound (research is done).
      // Mark research.needed=false so postMessages doesn't re-attach the tool.
      const reviewPayload: Record<string, unknown> = {
        ...enrichedPayload,
        research: {
          ...((enrichedPayload.research as Record<string, unknown> | undefined) ?? {}),
          needed: false,
          mode: 'provided_current_facts',
          providedEvidence: true,
        },
      };
      const messages = buildReviewMessages(reviewPayload);
      const response = await this.postMessages(messages, REVIEW_MAX_TOKENS, REVIEW_TEMPERATURE, signal, reviewPayload);
      return finalizeReviewResultForPayload(parseModelJson(extractAnthropicText(response)), reviewPayload);
    }
    const messages = buildReviewMessages(payload);
    const response = await this.postMessages(messages, REVIEW_MAX_TOKENS, REVIEW_TEMPERATURE, signal, payload);
    return finalizeReviewResultForPayload(parseModelJson(extractAnthropicText(response)), payload);
  }

  /**
   * Research pass — separate LLM call with web search bound. Captures the model's research
   * summary + citations and returns the original payload with `context.researchEvidence`
   * populated, ready for the structured review pass to consume.
   *
   * Failures here are non-fatal: if the research pass errors, we return the original payload
   * unchanged and let the review pass do its own (single-pass) thing as before.
   */
  private async runResearchPass(
    payload: Record<string, unknown>,
    signal: AbortSignal | undefined,
  ): Promise<Record<string, unknown>> {
    try {
      const messages = buildResearchMessages(payload, researchTargetsForPayload(payload));
      // Force research.needed=true on the inner payload so postMessages attaches web_search.
      const innerPayload: Record<string, unknown> = { ...payload, research: { ...(payload.research as Record<string, unknown> | undefined), needed: true } };
      const response = await this.postMessages(messages, REVIEW_MAX_TOKENS, REVIEW_TEMPERATURE, signal, innerPayload);
      const rawSummary = extractAnthropicText(response).trim();
      const rawCitations = extractAnthropicCitations(response);
      const instruction = extractInstructionText(payload);
      const filteredCitations = filterLowAuthorityCitations(rawCitations, instruction);

      // If filtering dropped every citation AND the question is a pricing question,
      // suppress the summary too — better to let the structured review fall through to
      // needs_input than surface a stale answer from a discarded blog post.
      const dropped = rawCitations.length > 0 && filteredCitations.length === 0;
      const pricingQuestion = isPricingInstruction(instruction);
      const summary = (dropped && pricingQuestion)
        ? 'Current pricing could not be verified against an official source. Ask the user to confirm the plan name and price.'
        : (rawSummary || 'Research ran but produced no summary text.');

      if (filteredCitations.length === 0 && !rawSummary) return payload;
      const researchEvidence = {
        status: 'checked' as const,
        required: true,
        provider: 'Anthropic',
        checkedAt: new Date().toISOString(),
        summary,
        sources: filteredCitations.map((c) => ({ url: c.url, ...(c.title ? { title: c.title } : {}) })),
        sourcePolicy: 'Prefer official sources for prices and product facts. When a vendor publishes a plan page, use it as primary. Reject blog subdomains (blog.*, news.*) as primary sources for current pricing. Cite each fact with the official URL.',
      };
      const prevContext = (payload.context && typeof payload.context === 'object' && !Array.isArray(payload.context))
        ? payload.context as Record<string, unknown>
        : {};
      return { ...payload, context: { ...prevContext, researchEvidence } };
    } catch {
      // Research pass failure must not block the review — return the payload unchanged.
      return payload;
    }
  }

  async ask(payload: Record<string, unknown>, signal?: AbortSignal): Promise<Record<string, unknown>> {
    const messages = buildAskMessages(payload);
    const response = await this.postMessages(messages, ASK_MAX_TOKENS, ASK_TEMPERATURE, signal, payload);
    const text = extractAnthropicText(response);
    if (text.trim().length === 0) {
      throw new ProviderHttpError(
        PROVIDER_ERROR_CODES.INVALID_RESPONSE,
        'Provider response had no answer text.',
      );
    }
    return { output_text: text };
  }

  private async postMessages(
    messages: DeviceAgentMessages,
    maxTokens: number,
    temperature: number,
    signal: AbortSignal | undefined,
    payload: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    const apiKey = (this.config.apiKey ?? '').trim();
    assertApiKeyHeaderSafe(apiKey);

    const apiFormat = isOpenRouterConfig(this.config) ? 'openai-compatible' : 'anthropic';
    const baseUrl = normalizeBaseUrl(this.config.baseUrl, apiFormat);
    const url = `${baseUrl}/messages`;

    const body: Record<string, unknown> = {
      model: this.config.model.trim(),
      max_tokens: maxTokens,
      system: messages.system,
      messages: [{ role: 'user', content: messages.userContent }],
      temperature,
      ...(researchNeeded(payload) ? { tools: [webSearchToolForConfig(this.config, payload)] } : {}),
    };

    const headers: Record<string, string> = isOpenRouterConfig(this.config)
      ? {
        Authorization: `Bearer ${apiKey}`,
        'X-OpenRouter-Metadata': 'enabled',
        ...openRouterAttributionHeaders(true),
      }
      : {
        'x-api-key': apiKey,
        'anthropic-version': ANTHROPIC_VERSION,
        'anthropic-dangerous-direct-browser-access': 'true',
      };

    const response = await this.http.postJson(url, headers, JSON.stringify(body), signal);
    const errorCode = mapHttpStatusToErrorCode(response.status);
    if (errorCode !== null) {
      throw new ProviderHttpError(errorCode, composeErrorMessage(response.status, response.body));
    }
    let parsed: unknown;
    try {
      parsed = JSON.parse(response.body);
    } catch {
      throw new ProviderHttpError(
        PROVIDER_ERROR_CODES.INVALID_RESPONSE,
        'Provider response was not valid JSON.',
      );
    }
    if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
      throw new ProviderHttpError(
        PROVIDER_ERROR_CODES.INVALID_RESPONSE,
        'Provider response was not valid JSON.',
      );
    }
    return parsed as Record<string, unknown>;
  }
}

function researchNeeded(payload: Record<string, unknown>): boolean {
  const research = payload.research;
  return Boolean(research && typeof research === 'object' && !Array.isArray(research) && (research as Record<string, unknown>).needed === true);
}

function extractInstructionText(payload: Record<string, unknown>): string {
  const direct = typeof payload.instruction === 'string' ? payload.instruction : '';
  if (direct.length > 0) return direct;
  const userPrompt = typeof payload.userPrompt === 'string' ? payload.userPrompt : '';
  if (userPrompt.length > 0) return userPrompt;
  const question = typeof payload.question === 'string' ? payload.question : '';
  return question;
}

function researchMaxUses(payload: Record<string, unknown>): number {
  const research = payload.research;
  if (!research || typeof research !== 'object' || Array.isArray(research)) return 3;
  const value = (research as Record<string, unknown>).maxSearches;
  return typeof value === 'number' && Number.isFinite(value) && value > 0
    ? Math.max(1, Math.min(Math.floor(value), 5))
    : 3;
}

function isOpenRouterConfig(config: RuntimeConfig): boolean {
  return config.provider.trim().toLowerCase() === 'openrouter' || (config.baseUrl ?? '').includes('openrouter.ai');
}

function webSearchToolForConfig(config: RuntimeConfig, payload: Record<string, unknown>): Record<string, unknown> {
  return isOpenRouterConfig(config) ? openRouterWebSearchTool() : anthropicWebSearchTool(payload);
}

function openRouterWebSearchTool(): Record<string, unknown> {
  return {
    type: 'openrouter:web_search',
    parameters: {
      engine: 'auto',
      max_total_results: 3,
      user_location: {
        type: 'approximate',
        country: 'US',
        timezone: 'America/Los_Angeles',
      },
    },
  };
}

function anthropicWebSearchTool(payload: Record<string, unknown>): Record<string, unknown> {
  return {
    type: 'web_search_20250305',
    name: 'web_search',
    max_uses: researchMaxUses(payload),
    user_location: {
      type: 'approximate',
      country: 'US',
      timezone: 'America/Los_Angeles',
    },
  };
}
